### PR TITLE
cloud: add operation create

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: coap-gateway/coapconv/coapconv.go [hunk 1]
```diff
@@ -268,3 +269,38 @@ func MakeRetrieveResourceRequest(deviceID, href string, resourceInterface, corre
 		CommandMetadata:   &metadata,
 	}
 }
+
+func MakeConfirmResourceCreateRequest(deviceID, href string, correlationId string, authCtx *commands.AuthorizationContext, connectionID string, req *pool.Message) commands.ConfirmResourceCreateRequest {
```
Comment (ondrejtomcik) [inline coap-gateway/coapconv/coapconv.go]: Why are you not working here with `commands.ResourceId` ?

File: coap-gateway/service/client.go [hunk 2]
```diff
@@ -789,3 +790,89 @@ func (client *Client) unpublishResources(ctx context.Context, resourceIDs []pbRA
 
 	client.unobserveResources(ctx, resourceIDs, rscsUnpublished)
 }
+
+func (client *Client) sendErrorConfirmResourceCreate(deviceID, href, userID, correlationID string, authCtx *commands.AuthorizationContext, code codes.Code, errToSend error) {
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: Same here. Use `commands.ResourceId`

File: grpc-gateway/client/client.go [hunk 3]
```diff
@@ -298,7 +297,7 @@ func (c *Client) RetrieveResourcesByResourceIDs(
 	tc := make(map[string]func(pb.ResourceValue), len(resourceIDsCallbacks))
 	resourceIDs := make([]*pb.ResourceId, 0, len(resourceIDsCallbacks))
 	for _, c := range resourceIDsCallbacks {
-		tc[utils.MakeResourceID(c.ResourceID.DeviceId, c.ResourceID.Href)] = c.Callback
+		tc[c.ResourceID.GetDeviceId()+c.ResourceID.GetHref()] = c.Callback
```
Comment (ondrejtomcik) [inline grpc-gateway/client/client.go:300]: Why are you not using UUID here? `c.ResourceID.ToUUID()`

File: resource-aggregate/.plantuml [hunk 4]
```diff
@@ -146,4 +146,37 @@ deactivate ResourceAggregate
...
+Server -> Gateway : OK
+deactivate Server
+Gateway -> ResourceAggregate : ConfirmResourceCreateRequest
+activate ResourceAggregate
+ResourceAggregate -> Gateway : ConfirmResourceCreateResponse
+deactivate Gateway
+ResourceAggregate --> Bus : ResourceCreated
+deactivate ResourceAggregate
+Bus --> Client : ResourceCreated
+deactivate Client
```
Comment (ondrejtomcik) [inline resource-aggregate/.plantuml:183]: Please extend it with publishing which follows . Same for unpublish and delete.
Thanks

File: resource-aggregate/commander/commander.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+	"fmt"
+
+	"github.com/plgd-dev/cloud/resource-aggregate/commands"
+	"github.com/plgd-dev/cloud/resource-aggregate/cqrs/eventbus"
+	"github.com/plgd-dev/cloud/resource-aggregate/cqrs/utils"
+	"github.com/plgd-dev/cloud/resource-aggregate/events"
+	"github.com/plgd-dev/cloud/resource-aggregate/service"
+)
+
+// Commander provides sync commands.
```
Comment (ondrejtomcik) [inline resource-aggregate/commander/commander.go]: What are sync commands?

File: resource-aggregate/commander/commander.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
+package commander
```
Comment (ondrejtomcik) [inline resource-aggregate/commander/commander.go]: Commander isn't the best name.It's something completely different. Describe its purpose and we will find a better name.
https://slovniky.lingea.sk/anglicko-slovensky/commander

File: resource-aggregate/pb/commands.proto [hunk 7]
```diff
@@ -314,4 +314,69 @@ message ConfirmResourceDeleteRequest {
...
+     │OCF Server│          │Gateway│                   │Resource Aggregate│          │Event Bus│          │Client│
+     └──────────┘          └───────┘                   └──────────────────┘          └─────────┘          └──────┘
+******************************************************************************************************************************************************/
+
+message CreateResourceRequest {
+    AuthorizationContext authorization_context = 1;
+    ResourceId resource_id = 2;
+    string correlation_id = 3;
+    Content content = 4;
+    CommandMetadata command_metadata = 100;
```
Comment (ondrejtomcik) [inline resource-aggregate/pb/commands.proto:380]: Why is there 100?

File: resource-aggregate/pb/commands.proto [hunk 8]
```diff
@@ -314,4 +314,69 @@ message ConfirmResourceDeleteRequest {
...
+message CreateResourceResponse {
+    AuditContext audit_context = 1;
+}
+
+message ConfirmResourceCreateRequest {
+    AuthorizationContext authorization_context = 1;
+    ResourceId resource_id = 2;
+    string correlation_id = 3;
+    Status status = 4;
+    Content content = 5;
```
Comment (ondrejtomcik) [inline resource-aggregate/pb/commands.proto:392]: Is there any content returned from the device? Where is the href of newly created resource available?

File: coap-gateway/coapconv/httpCode2CoapCode.go [hunk 9]
```diff
@@ -21,8 +21,6 @@ func HttpCode2CoapCode(statusCode int, method coapCodes.Code) coapCodes.Code {
 			return coapCodes.Changed
 		case coapCodes.GET:
 			return coapCodes.Content
-		case coapCodes.PUT:
```
Comment (ondrejtomcik) [inline coap-gateway/coapconv/httpCode2CoapCode.go:24]: Why did you remove `PUT` ?

File: resource-aggregate/commander/commander.go [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+func newUpdateHandler(correlationID string) *updateHandler {
+	return &updateHandler{
+		correlationID: correlationID,
+		res:           make(chan *events.ResourceUpdated, 1),
+	}
+}
+
+func (h *updateHandler) Handle(ctx context.Context, iter eventbus.Iter) (err error) {
```
Comment (ondrejtomcik) [inline resource-aggregate/commander/commander.go]: Is this supposed to be just a helper so each gateway is not implementing it on it's own? Why it's not containing all commands RA support?

File: resource-aggregate/service/aggregate_test.go [hunk 11]
```diff
@@ -496,6 +515,26 @@ func testMakeConfirmResourceDeleteRequest(deviceID, href, correlationID string)
 	return &r
 }
 
+func testMakeConfirmResourceCreateRequest(deviceID, href, correlationID string) *commands.ConfirmResourceCreateRequest {
```
Comment (ondrejtomcik) [inline resource-aggregate/service/aggregate_test.go]: Please move right under the `testMakeCreateResourceRequest`.

File: resource-aggregate/events/resourceStateSnapshotTaken.go [hunk 12]
```diff
@@ -381,6 +411,46 @@ func (e *ResourceStateSnapshotTaken) HandleCommand(ctx context.Context, cmd aggr
 			return nil, err
 		}
 		return []eventstore.Event{&rc}, nil
+	case *commands.CreateResourceRequest:
+		if req.GetCommandMetadata() == nil {
+			return nil, status.Errorf(codes.InvalidArgument, errInvalidCommandMetadata)
+		}
+
+		em := MakeEventMeta(req.GetCommandMetadata().GetConnectionId(), req.GetCommandMetadata().GetSequence(), newVersion)
```
Comment (ondrejtomcik) [inline resource-aggregate/events/resourceStateSnapshotTaken.go:419]: I would propose to validate if this resource is of type `oic.wk.collection`

**Review by ondrejtomcik — CHANGES_REQUESTED** [review]

**Review by ondrejtomcik — APPROVED** [review]

File: resource-aggregate/pb/commands.proto (hunk 8 above)
Comment (jkralik) [reply resource-aggregate/pb/commands.proto:392]: In the content.

File: coap-gateway/coapconv/httpCode2CoapCode.go (hunk 9 above)
Comment (jkralik) [reply coap-gateway/coapconv/httpCode2CoapCode.go:24]: PUT is not supported by OCF.

File: resource-aggregate/commander/commander.go (hunk 10 above)
Comment (jkralik) [reply resource-aggregate/commander/commander.go]: Only these commands waits for events from the target device.

File: resource-aggregate/events/resourceStateSnapshotTaken.go (hunk 12 above)
Comment (jkralik) [reply resource-aggregate/events/resourceStateSnapshotTaken.go:419]: At the resource with don't know resource type because this values are stored at /plgd/res resource.

File: resource-aggregate/pb/commands.proto (hunk 7 above)
Comment (jkralik) [reply resource-aggregate/pb/commands.proto:380]: place for other properties
